extension_service/datastore/providers/firestore.py

```diff
@@ -0,0 +1,270 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+from typing import Any, Dict, Literal, Optional
+
+import firebase_admin
+from firebase_admin import credentials, firestore_async
+from google.cloud import firestore
+from google.cloud.firestore_v1.base_query import FieldFilter
+from pydantic import BaseModel
+
+import models
+
+from .. import datastore
+
+
+class Config(BaseModel, datastore.AbstractConfig):
+    kind: Literal["firestore"]
+    projectId: str
+    serviceAccountId: str
+
+
+class Client(datastore.Client[Config]):
+    __client: firestore.AsyncClient
+
+    @datastore.classproperty
+    def kind(cls):
+        return "firestore"
+
+    def __init__(self, client: firestore.AsyncClient):
+        self.__client = client
+
+    @classmethod
+    async def create(cls, config: Config) -> "Client":
+        cred = credentials.ApplicationDefault()
+        firebase_admin.initialize_app(
+            cred,
+            options={
+                "projectId": config.projectId,
+                "serviceAccountId": config.serviceAccountId,
+            },
+        )
+        return cls(firestore.AsyncClient(project=config.projectId))
+
+    async def initialize_data(
+        self,
+        airports: list[models.Airport],
+        amenities: list[models.Amenity],
+        flights: list[models.Flight],
+    ) -> None:
+        async def delete_collection(coll_ref, batch_size=400):
+            docs = coll_ref.limit(batch_size).stream()
+            deleted = 0
+
+            async for doc in docs:
+                await doc.reference.delete()
+                deleted = deleted + 1
+
+            if deleted >= batch_size:
+                return await delete_collection(coll_ref, batch_size)
```

Comment:
I don't understand what the batch_size is or why it's set to 400 (or why it's a parameter). Why are we counting and returning delete_collection here? Can we include some comments on why this is necessary?

Reply:
When deleting a collection in firestore, we need to retrieve all documents and delete each one. It is recommended to delete them in smaller batches to avoid `out-of-memory` error according to https://firebase.google.com/docs/firestore/manage-data/delete-data#collections. Since a batched write can contain maximum 500 operations, `batch_size` should be < 500 and I set it to 400 to be safe. It doesn't have to be a parameter though, I could just use a hardcoded value if that looks better.

Reply:
After discussion with Daniel, we determined it is unnecessary to delete recursively in batches. Instead we iteratively create task for each delete, and await them together with `asyncio.gather()`